src/pybind/mgr/prometheus/module.py

```diff
@@ -505,6 +508,146 @@ def get_num_objects(self):
             stat = 'num_objects_{}'.format(obj)
             self.metrics[stat].set(pg_sum[stat])
 
+    def get_rbd_stats(self):
+        COUNTER_DESCRIPTORS = {
+            'write_ops' : {'type' : self.PERFCOUNTER_COUNTER,
+                           'desc' : 'RBD image writes count'},
+            'read_ops' : {'type' : self.PERFCOUNTER_COUNTER,
+                          'desc' : 'RBD image reads count'},
+            'write_bytes' : {'type' : self.PERFCOUNTER_LONGRUNAVG,
+                             'desc' : 'RBD image bytes written'},
+            'read_bytes' : {'type' : self.PERFCOUNTER_LONGRUNAVG,
+                            'desc' : 'RBD image bytes read'},
+            'write_latency' : {'type' : self.PERFCOUNTER_LONGRUNAVG,
+                               'desc' : 'RBD image writes latency (msec)'},
+            'read_latency' : {'type' : self.PERFCOUNTER_LONGRUNAVG,
+                              'desc' : 'RBD image reads latency (msec)'},
+        }
+
+        pools = self.get_localized_config('rbd_stats_pools', '').split()
+
+        for pool_id, pool in self.rbd_stats['pools'].items():
+            if pool['name'] not in pools:
+                del self.rbd_stats['pools'][pool_id]
+
+        for pool_name in pools:
+            try:
+                pool_id = self.rados.pool_lookup(pool_name)
+                with self.rados.open_ioctx(pool_name) as ioctx:
+                    if pool_id not in self.rbd_stats['pools']:
+                        self.rbd_stats['pools'][pool_id] = {'images' : {}}
+                    pool = self.rbd_stats['pools'][pool_id]
+                    pool['name'] = pool_name
+                    images = {}
+                    for image_meta in RBD().list2(ioctx):
```

Comment:
It's going to look-up all the images in the pools on each iteration of the prometheus status gathering? It would be great if this could be caches and only refreshed as needed (i.e. unknown image id in the results)

Reply:
Not exactly on each iteration -- if it is gathered more frequently than `scrape_interval` then just a cached result is returned, but `scrape_interval` is small by default (5 sec). So yes, I think it will usually fired on every prometheus gathering.

I think I can't just refresh only  if an unknown image id is found in the result. The module reports stats for all images, including inactive. So if an image is added but not used yet, the stats will not be produced. And if an image is removed the module will continue to return its stats.

I was planning to improve this later, but I can do in this PR.  I think I will add an additional param `rbd_stats_pool_refresh_interval`, setting it to 5 minutes, so it will not refresh on every gathering. And will also add force refresh when an unknow image id found in the result.

Reply:
Sounds good